Multiple active gateway enhancement proposal (#69)
submariner-io/enhancements: submariner/multiple-active-gateways.md

# Multiple Active Gateways

## Summary

NOTE: The design details are still a work in progress, so this is currently a draft proposal.

Submariner currently only allows a single gateway to be active at any one time in a given cluster.

> This enhancement proposes that there be an option to allow more than one gateway to be active in

Reviewer: In this new design, will we be using the same submariner gateway label to mark a node[/s] as Gateway nodes?

And all the labelled Gateway nodes will automatically be part of Active/Active gateways?

Author: That is the plan.  We think it's simpler than having some nodes be active gateways, while others are backups.  However, we think enabling the later could be useful.  Some options could include explicitly labeling active/backup, or have a setting that allows us to configure N active gateways, and then have the first N go active, and any additional gateways would be passive unless and until one of the active gateways fails.

a given cluster to enhance both performance and resiliency. This feature is sometimes referred to as

> “active-active gateways”; however, the goal is to allow two or more actove gateways to be used.

Reviewer 2: /s/actove/active

## Proposal

The proposal is to optionally allow multiple gateways to be active at the same time in a given cluster.

When only a single active gateway is used, all inter-cluster traffic must flow through that one gateway,

> which creates a performance bottleneck. Enabling more than one node to be used as a gateway can improve

Reviewer 2: Maybe go in a little more about how this is a performance bottleneck specifically with encryption engines? I.e All encryption is locked to one core etc.

Author: Ok.

Reviewer 3: It seems that the painful points in the current solution are performance bottleneck and slow failover time.

While the proposed solution addresses the performance bottleneck, it complicates the datapath handling both intra-cluster and inter-cluster. 

Are there other alternatives (e.g: setting more resources somehow to vpn pod, etc) to mitigate the performance bottleneck?

Do we know how critical is the failover timing? the maximum time for GW failover? 
Is it possible to reduce GW failover timing with Active/SB architecture?  (Maybe we can use VIP to represent the active GW? )

Author: Improving the recovery time isn't the main objective.  So, I think of it as more of a bonus.  I'm sure there is room for improvement in the current design/implementation without going to MAG.

the performance linearly in the number of gateways used.

> Additionally, when the active gateway fails in the current implementation, work must be done to switch

Reviewer: Can you please describe how we will figure of when the active gateway fails?

PS: In the current design, we use the endpoint objects to identify the active gateway nodes. When a new node becomes the active gateway node, it deletes the old/stale endpoint.

Author: I don't have an answer right now, but we'd know this needs to be worked out.  We can add it to TODO list.

Reviewer 4: This seems like something we should design for as it could get quite complicated.

Author: If there is a decision to move forward with this enhancement, then we can work out the details.

over to the new gateway including the establishment of new tunnels (both intra-cluster and inter-cluster)
and new routes, which takes time. With the proposed design, It is expected that the fail-over times can
be significantly improved if additional gateways are operational. Furthermore, it is expected that
traffic flowing through operational gateways will continue to flow while the traffic from the failed
gateway is redirected to the operational gateways. Both properties reduce the impact of a failure.

## Design Details

### Single Gateway

> In the current single gateway Submariner implementation, intra-cluster VXLAN tunnels are established by

Reviewer 4: This is only true for `iptables` based CNIs, other CNIs don't necessarily need these internal tunnels.

Author: For which CNIs does Submariner not use a tunnel for inter-cluster traffic between non-gateway and gateway nodes?

Reviewer 4: I see this listed under TODO as "* Support non-kubeproxy route agent handlers."

the non-gateway nodes to the gateway node, and each gateway establishes an inter-cluster tunnel to each
other gateway in the clusterset (with the tunnel type determined by the cable driver in use) as shown
below.

![SingleGateway](./images/single-gateway.png)

### Multiple Active Gateway Design

The proposal for supporting multiple active gateway nodes is shown in the following diagram and described below.

![MultiGateway](./images/multi-gateway.png)

> * Each non-gateway node will establish a VXLAN tunnel to each gateway node.

Reviewer 2: Do we have any understanding of the performance impact with the configuration of so many more tunnels per clusterset?

Author: This is a good point and something that should be covered.  We don't have hard numbers.  However, the goal and expectation are that this will improve forwarding performance. Initial setup could take a little longer for each new endpoint (peer cluster), but it shouldn't be too big of a hit.  However, this could increase the total number of tunnels for a clusterset significantly.  For N nodes, we currently have N(N-1) tunnels. If there are M gateways on each node, we multiply that by 2M^2.  So, I'm thinking that for large deployments, we'd only want to use multiple gateways in a set of high speed "core" clusters, and then connect "leaf" clusters in with fewer connections to a subset of core clusters.  However, this will require that the gateways have "transit node" functionality and that we use dynamic routing, so that's for a future enhancement :-).

Author: So, in summary, I do have some scalability concerns for large deployments if we don't also make other enhancements. We'll have to analyze the impact.

> * The non-gateway nodes will load-balance flows between the active gateway nodes.

Reviewer 2: Following the POC It will be good to have an "implementation" section going into the details of these bits

Author: Agreed. I was mainly trying to get a high-level description documented here. We should add more details once we have a few more things figured out.

* Each gateway node will establish a tunnel to each peer gateway node.

> * The gateway nodes will load-balance flows between the active peer gateway nodes.

Reviewer 4: What does this mean?

Author: This has been expanded on in the updated proposal.

> * Load balancing will be performed using Linux resilient nexthop groups.

Reviewer: Currently the Active Submariner Gateway pod is responsible for couple of things like
1. Creating local endpoint and cluster objects 
2. Syncing the endpoints/cluster objects with the Broker Cluster
3. Healthchecks with remote cluster gateway nodes
4. Performing NAT discovery to identify if remote cluster is reachable over private or public-ip in the endpoint.
5. Gateway object syncer.
6. Supporting metrics

Since some of the above operations have to be done by only one component, can you please add notes on how the above operations will be handled in the new Active/Active architecture.

Author: @astoycos and @Billy99 are working on a POC of this now and we'll update this proposal when we work it out.

Author: This has been described in the updated proposal.

> * Load balancing will be performed using Linux resilient nexthop groups.

Reviewer 4: A link to some documentation, or even a brief explanation, of "Linux resilient nexthop groups" will help understand what's expected to be happening

Author: Done.

### Multiple Active Gateway Packet Flow

An example packet flow is shown in the following diagram and described below.

![MultiGateway](./images/multi-gateway-flow.png)

1. Node 1-1

>     1. Intercept packet destined for global service

Reviewer 4: Is there an implicit assumption that globalnet is deployed? We need to support deployments both with and without it

    2. Load balances flow to gateway 1-1
    3. Hash ensures remaining packets in the connection also go to gateway 1-1
2. Gateway 1-1
    1. Determines packet is destined for Cluster 2 based on dest Globalnet IP
    2. Load balances flow to Gateway 2-2

>     3. Perform SNAT from local source/port to one of the Globalnet address/ports for gateway 1-1

Reviewer 4: This needs to conform with Globalnet 2.0

Author: It does.

    4. Hash ensures that remaining packets in the flow also go to Gateway 2-2
3. Gateway 2-2

>     1. The packet is handed off to the local service load balancer, which performs necessary NAT

Reviewer 4: Since you mentioned globalnet, it should also perform NAT here for the destination, before its further processed

Author: The details have been included in the updated proposal.

operations and forwards the packet to chosen endpoint on Node 2-1.
4. Node 2-1
    1. Send response back to Gateway 2-2 (through service load balancer)

>     2. Note: it should be okay if the return path uses a different gateway, but I think the cluster

Reviewer 4: How do you make sure in this case it routes back through the same gateway? Or is it automatic?

Author: The details have been included in the updated proposal.

load balancer will return it to the same node in order to maintain the NAT state.
5. Gateway 2-2
    1. Forward back to Gateway 1-1 over the appropriate tunnel

>     2. Note: It is important that Gateway 2-2 forwards the packet directly to Gateway 1-1 and not

Reviewer 4: Again its not clear hos you determine to send it back to the same gateway?

Author: The details have been included in the updated proposal.

via the nexthop load balancer which may send it to Gateway 1-2. This can be accomplished by using host
routes for the gateway IP addresses that are higher priority than the routes for the cluster Globalnet
CIDRs which would use the nexthop load balancing.
6. Gateway 1-1
    1. Perform DNAT from the global address back to the local address
    2. Send to the pod on Node 1-1 via local CNI.

## Notes and Assumptions

> * This design assumes that Globalnet is being used. It may be possible to make it work without Globalnet if

Reviewer 4: It's essential this works without globalnet.

Moreover, globalnet is an optional layer on top of core Submariner so I'd say to get it working without globalnet is higher priority.

Author: The design and POC now support non-Globalnet.

necessary, but step 5 relies on the fact that the dest IP will be the address for the correct gateway.
Without Globalnet, we’d still need to do the same type of SNAT operation at step 2 using local addresses
which isn’t currently done by Submariner.

* The initial POC will use standard external IP addresses for gateways, but load balancer mode is
a high priority for the final implementation.

* The initial design and POC is focused on non-ovn-kubernetes CNIs, but ovn-kubernetes must also be supported.

* There will be a config option to enable multiple active gateways, and when enabled, all nodes labeled with
“submariner.io/gateway=true” will become active. It is for future study whether we should support an option
where we support more than one active gateway as well as additional backup inactive gateways.

> ## Plan

Reviewer 4: I don't see any alternatives, nor do I see any detailed design (except a theoretical example packet flow)

Author: The details have been included in the updated proposal.

    Routing Table 100: 10.1.0.0/16 src 10.1.128.0 metric 100 pervasive
                       nexthop via 241.18.0.6 dev vxlan-tunnel weight 1
                       nexthop via 241.18.0.8 dev vxlan-tunnel weight 1
    ```

    `SRC: 100.2.0.132  DST: 10.1.128.4`

6. Traffic is received on the inter-cluster VxLAN Tunnel on either **GW-1** or **GW-2**.

   a. The traffic is accepted and forwarded to the original originating pod.

    `SRC: 100.2.0.132  DST: 10.1.128.4`

   b. The following rule includes `-m conntrack --ctstate DNAT` so that the SNAT step described in 3-b is skipped for
response packets.

   ```shell
    SKIPPED:  -A SUBMARINER-POSTROUTING -i vxlan-tunnel -o weave -m conntrack --ctstate DNAT -j SNAT --to-source 10.1.128.0
    ```

   NOTE: Weave performs connection tracking on flows and only allows responses to TCP sessions with state.  The state for
the given flow exists on **GW-2**, but not on **GW-1**. Because of this, it is necessary to add the following rule to allow
the response packets through if they happen to be returned to a gateway other than the one that the outbound flow uses:

   ```shell
    -A FORWARD -d 10.1.0.0/16 -i vxlan-tunnel -j ACCEPT
    ```

>    An alternative approach is to SNAT packets using egress IPs in a way similar to what is done for Globalnet with

Reviewer 5: ```suggestion
**NOTE:** An alternative approach is to SNAT packets using egress IPs in a way similar to what is done for Globalnet with
```

tracking state for step 3-a only exists on this node. To do this, the traffic is SNAT’ed to the IP address of the Node’s main
CNI interface so that traffic returns via the same gateway. For this setup the Weave CNI is used so service DNAT is handled by
kubeproxy. The IPtables rule looks something like:

   ```shell
    A SUBMARINER-POSTROUTING -i vxlan-tunnel -o weave -m conntrack --ctstate DNAT -j SNAT --to-source 10.2.192.0
    ```

    `SRC: 10.2.192.0  DST: 10.0.128.5`

   c. Routed to the Dst node by the CNI.

    `SRC: 10.2.192.0  DST: 10.0.128.5`

4. Traffic is delivered to the destination pod via the CNI, and sent back to the correct local GW node (**GW-4**) via the
CNI due to the SNAT completed in step 3-b.

    `SRC: 10.0.128.5  DST: 10.2.192.0`

5. Traffic arrives back at the same GW node (**GW-4**) it entered on:

   a. Traffic is DNAT'ed (reverse the SNAT in 3-b) back to one of the Original GW node’s `clusterglobalEgressIP`.

    `SRC: 10.0.128.5 DST: 242.254.1.1`

   b. Traffic is SNAT'ed (reverse the DNAT in 3-a) back to external IP of the k8s service mirrored by submariner.

    `SRC: 242.254.2.254  DST: 242.254.1.1`

>    c. Routed back to the Original gateway using specific routes have been added for ClusterGlobalEgressIPs that take

Reviewer 5: ```suggestion
   c. Routed back to the Original gateway using specific routes that have been added for ClusterGlobalEgressIPs that take
```

precedence over the multipath routes that are used for the larger Globalnet CIDR.

   ```shell
    Routing Table 100:   242.254.1.10 via 241.18.0.8 dev vxlan-tunnel metric 98
                         …
                         242.254.1.16 via 241.18.0.8 dev vxlan-tunnel metric 98
    ```

    `SRC: 242.254.2.254  DST: 242.254.1.1`

6. Traffic is DNAT’ed back to the original pod IP and sent back via the CNI.

    `SRC: 242.254.2.254  DST: 10.1.128.4`

## Data Model Changes

### Enabling Multiple Active Gateways Feature

Multiple Active Gateways is an optional feature that can be enabled when a cluster is joined to the broker.
The user can indicate whether they wish the feature to be enabled or not using a new flag added to the

> subctl join` command.

Reviewer 5: ```suggestion
`subctl join` command.
```

// +k8s:openapi-gen=true
type SubmarinerSpec struct {
    Broker string `json:"broker"`
    :
    MultiActiveGatewayEnabled bool `json:"multiActiveGatewayEnabled,omitempty"`
    :
}
```

The flag will need to be passed to the `submariner-gateway` and `submariner-globalnet` pods when they are
started. Within each pod, the flag will be propagated to multiple controllers.

### Non-Globalnet

No data-model changes specific to the non-Globalnet deployment.

### Globalnet

#### Globalnet ClusterGlobalEgressIps Instancing

Currently, there is one instance of the `clusterglobalegressips` CRD object per cluster. This object is a set of
globally unique IP Addresses (across all clusters managed by Submariner) one of which is used on the Gateway node
as the IP Address to SNAT the source address on egress packets. Currently, the one instance of the
`clusterglobalegressips` CRD object uses the default name of `cluster-egress.submariner.io`.

With Multiple Active Gateways, there will need to be a `clusterglobalegressips` CRD object per Gateway, each of
which is a globally unique set of IPs. To adjust for multiple instances, the node name is prepended to the existing
constant instance name.

> *(TBD - Do we need to keep non-MAG name of object as is and only prepend node name if MAG enbable? POC always

Reviewer 5: ```suggestion
*(TBD - Do we need to keep non-Multiple Active Gateway naming of the object as is and only prepend node name if Multiple Active Gateway is enabled for upgrade reasons? POC always
```

> *(TBD - Do we need to keep non-Multiple Active Gateway naming of the object as is and only prepend node name if Multiple Active Gateway is enabled for upgrade reasons? POC always

Author: ```suggestion
*(TBD - Do we need to keep non-Multiple Active Gateway naming of the object as is and only prepend node 
name if Multiple Active Gateway is enabled for upgrade reasons? POC always
```

semaphore unchanged.

With Multiple Active Gateways enabled, there is no concept of a stand-by gateway. There are only active gateways.
With Multiple Active Gateways disabled, the code functions as before with one active gateway and all other
gateway pod instances waiting on the leadership semaphore to be released.

##### Move Globalnet to a Deployment from a Daemonset

###### Background

Both the Gateway and Globalnet pods are launched as a DaemonSet, where pods are scheduled on any node with the
label `submariner.io/gateway=true`. As mentioned above, the Gateway pods run based on the leadership semaphore.
The Globalnet pods on the other hand run in either a Gateway (running on the active Gateway) or Non-Gateway
(running on the standby Gateway) mode. In both modes, the Gateway Monitor Controller watches all Submariner
Endpoints and for remote Endpoints, updates the IP Table rules to mark packets destined for the remote cluster.

In the Gateway mode case, a lot more is done. The IP Pool is created and all the other controllers are started.
The IP Pool is used to reserve IP Addresses used by Globalnet (ClusterGlobalEgressIP, GlobalEgressIP, and

> GlobalIngressIP). Whenever IP Addresses are reserved or released by one of the controllers, associated IP Table

Reviewer 5: ```suggestion
GlobalIngressIP). Whenever IP Addresses are reserved or released by one of the controllers, associated datapath
```

With multiple gateways running in a cluster, running multiple Globalnet pods is problematic. The primary issue
is that with multiple Globalnet pods there are multiple IP allocators, each running from the same IP Pool. One
option is to move the IP Allocator out of Globalnet and place it behind a local service, allowing multiple
Globalnet pods to request IPs. However, only ClusterGlobalEgressIP is per node, the other two controllers
requesting IP allocation are GlobalEgressIP and GlobalIngressIP, which are per cluster.

The proposed solution is to move the Globalnet pod from being a Daemonset that mirrors the Gateway Daemonset
to a Kubernetes Deployment with a ReplicaSet of 1. With only one instance of the Globalnet pod, there will
only be one IP Allocator and one instance of each of the controllers that are requesting the IPs.

Since the Globalnet pod is not running on every node that needs local datapath updates, the Globalnet pod cannot
update the datapath rules as it currently does. The second part of the proposal is to move datapath rule management
out of the Globalnet pod and into the Gateway pod, which runs on each Gateway node. The controllers running in

> Globalnet pod will still manage IP allocation and the CRD object management, just remove the IP Table updates

Reviewer 5: ```suggestion
Globalnet pod will still manage IP allocation and the CRD object management, just remove the local datapath updates
```

from the following controllers in the Globalnet pod:

* Gateway Monitor Controller
* Node Controller
* Cluster Global Egress IP Controller
* Global Ingress IP Controller
* Global Egress IP Controller

Five new controllers with the same name will be added to the Gateway pod, which watch their respective CRD

> objects and modify the IP Tables as needed on their host. An alternative to this approach could be to add all of the

Reviewer 5: ```suggestion
objects and modify the local datapath as needed on their host. An alternative to this approach could be to add all of the
```

datapath functionality needed for globalnet to the RouteAgent, however this would result in a much higher level of
complexity for the RA, and a bigger re-write.

#### Expose ClusterGlobalEgressIPs to Remote Clusters

Referencing the [Multiple Active Gateway Packet Flow with Globalnet](#multiple-active-gateway-packet-flow-with-globalnet)
drawing, new routes need to be added to remote
cluster gateways (GW-4 in the drawing) that include the `ClusterGlobalEgressIPs` IP Addresses from each sending
node so that traffic is returned to the sending gateway (GW-1 in drawing) so that conntrack state of the packet
is maintained and the proper DNATing is performed (see Step 5-c). To enable these routes to be written, the
`ClusterGlobalEgressIPs` IP Addresses must be known on all gateways.

##### Option 1: Add ClusterGlobalEgressIP IP Addresses to Submariner EndPoint

The Submariner Endpoint CRD Object is currently already shared across clusters. A new field will be added

> to the existing `submariners` CRD object and used to store and disseminate the IP Addresses to the remote

Reviewer 5: ```suggestion
to the existing Submariner `endpoint` CRD object and used to store and disseminate the IP Addresses to the remote
```

clusters.

```go
type EndpointSpec struct {
    // +kubebuilder:validation:MaxLength=63
    // +kubebuilder:validation:MinLength=1
    ClusterID string `json:"cluster_id"`
    :
    // The list of allocated ClusterGlobaEgresslIPs.
    AllocatedIPs  []string          `json:"allocated_ips,omitempty"`
}
```

The advantage of using Submariner Endpoint is that it is already shared across clusters and instanced properly
for sharing the desired data.

> *NOTE:* This is how the POC was implemented.

Reviewer 5: ```suggestion
**NOTE:** This is how the POC was implemented.
```

The Cable Driver changes were similar to the Route Agent’s with some subtle differences.  The Cable Driver now needs to
create multipath routes in table 100 to all remote networks via all the GWs backing those remote CIDRs.

```shell
Routing Table 100:
<Remote Pod CIDR> proto static
      nexthop via <Remote GW 1> dev vxlan-tunnel weight 1
      nexthop via <Remote GW 2> dev vxlan-tunnel weight 1
<Remote Service CIDR> proto static
      nexthop via <Remote GW 1 > dev vxlan-tunnel weight 1
      nexthop via <Remote GW 2> dev vxlan-tunnel weight 1
```

To ensure this functionality the VXLAN Cable Driver’s route reconciliation logic was re-written to be more level driven as
well. Here, all the desired routes are calculated for every relevant remote endpoint event and the existing routes are simply
overridden with the total correct state.  This is slightly different from the re-written Route Agent reconciler, but could
be a bit less resource-intensive.  Ultimately we as a community should agree on the architecture and move all of the route
reconciliation logic to a common library.

### Iptables Architecture Changes

#### Route Agent Iptables Architecture Changes

> For the MAG functionality, two main IPtables changes were needed on every GW node.  The first was to ensure that K8’s service

Reviewer 5: ```suggestion
For the Multiple Active Gateway functionality, two main IPtables changes were needed on every GW node.  The first was to ensure that K8’s service
```

…
<Remote GW 1 SNAT IP X> via <Remote GW 1 IP> dev vxlan-tunnel metric 98
<Remote GW 2 SNAT IP 1> via <Remote GW 2 IP> dev vxlan-tunnel metric 98
…
<Remote GW 2 SNAT IP X> via <Remote GW 2 IP> dev vxlan-tunnel metric 98
```

## Scaling considerations

* Additional tunnels between clusters are required for this design.

## Notes and Assumptions

* The initial POC will use standard external IP addresses for gateways, but load balancer mode is
  a high priority for the final implementation.
* The initial design and POC is focused on CNIs that use the kubeproxy route agent handler that establishes VXLAN
  tunnels between worker nodes and gateways. Other options need to be investigated and supported.
* The multipath routes require an interface, so this approach won't work with IPsec Tunnel Mode.  IPsec Transport mode,
  proposed in [IPSec transport mode support](https://github.com/submariner-io/enhancements/pull/103) will be required for
  IPsec encryption.

## TODO

* Performance testing on POC code.
* Support non-kubeproxy route agent handlers.
* Support Loadbalancer mode.
* Support detection and handling of gateway failure and recovery.

## Links

* Source for images:

>   [Submariner-MAG-Drawings](https://docs.google.com/presentation/d/19W-0YJB0wDiEWDhs7MoLFgKEm5GkGs2g47jXMUnIrv0/edit#)

Reviewer 5: ```suggestion
  [Submariner-MAG-Drawings](https://docs.google.com/presentation/d/1PsxbBbzt9A9opO4bVxkJcLnzD6zAQTlKtK2_aq5SINs/edit#)
```
The previous link was the "Axon Data Plane Overview" slides. I added the link before I knew which we were going to use.

// +k8s:openapi-gen=true
type SubmarinerSpec struct {
    Broker string `json:"broker"`
    :
    MultiActiveGatewayEnabled bool `json:"multiActiveGatewayEnabled,omitempty"`
    :
}
```

The flag will need to be passed to the `submariner-gateway` and `submariner-globalnet` pods when they are
started. Within each pod, the flag will be propagated to multiple controllers.

### Non-Globalnet

No data-model changes specific to the non-Globalnet deployment.

### Globalnet

#### Globalnet ClusterGlobalEgressIps Instancing

Currently, there is one instance of the `clusterglobalegressips` CRD object per cluster. This object is a set of
globally unique IP Addresses (across all clusters managed by Submariner) one of which is used on the Gateway node
as the IP Address to SNAT the source address on egress packets. Currently, the one instance of the
`clusterglobalegressips` CRD object uses the default name of `cluster-egress.submariner.io`.

With Multiple Active Gateways, there will need to be a `clusterglobalegressips` CRD object per Gateway, each of
which is a globally unique set of IPs. To adjust for multiple instances, the node name is prepended to the existing
constant instance name.

> *(TBD - Do we need to keep non-Multiple Active Gateway naming of the object as is and only prepend node 

Author: ```suggestion
*(TBD - Do we need to keep non-Multiple Active Gateway naming of the object as is and only prepend node
```

> name if Multiple Active Gateway is enabled for upgrade reasons? POC always

Author: ```suggestion
name if Multiple Active Gateway is enabled for upgrade reasons? POC always
```

# Multiple Active Gateways

> ## Summary

Reviewer 5: ```suggestion
## Status

Deferred

## Summary
```

## Data Model Changes

### Enabling Multiple Active Gateways Feature

Multiple Active Gateways is an optional feature that can be enabled when a cluster is joined to the broker.
The user can indicate whether they wish the feature to be enabled or not using a new flag added to the
`subctl join` command.

```shell
subctl join … --multi-active-gateway=true
```

The flag will be processed by the `subctl join` command and used to populate a new field in the existing
`submariners` CRD object and used to store and disseminate the flag to the rest of the code. If not entered,
the new MultiActiveGatewayEnabled flag will default to false.

```go
// SubmarinerSpec defines the desired state of Submariner
// +k8s:openapi-gen=true
type SubmarinerSpec struct {
    Broker string `json:"broker"`
    :
    MultiActiveGatewayEnabled bool `json:"multiActiveGatewayEnabled,omitempty"`
    :
}
```

The flag will need to be passed to the `submariner-gateway` and `submariner-globalnet` pods when they are
started. Within each pod, the flag will be propagated to multiple controllers.

### Non-Globalnet

No data-model changes specific to the non-Globalnet deployment.

### Globalnet

> #### Globalnet ClusterGlobalEgressIps Instancing

Reviewer 5: ```suggestion
#### Globalnet ClusterGlobalEgressIps, GlobalEgressIps and GlobalIngressIps Instancing

**NOTE:** In the PoC, only the `clusterglobalegressips` CRD Object was updated as described below. However,  `globalegressips` and `globalingressips` CRD Objects will need similar changes.

```

only be one IP Allocator and one instance of each of the controllers that are requesting the IPs.

Since the Globalnet pod is not running on every node that needs local datapath updates, the Globalnet pod cannot
update the datapath rules as it currently does. The second part of the proposal is to move datapath rule management
out of the Globalnet pod and into the Gateway pod, which runs on each Gateway node. The controllers running in
Globalnet pod will still manage IP allocation and the CRD object management, just remove the local datapath updates
from the following controllers in the Globalnet pod:

* Gateway Monitor Controller
* Node Controller
* Cluster Global Egress IP Controller
* Global Ingress IP Controller
* Global Egress IP Controller

Five new controllers with the same name will be added to the Gateway pod, which watch their respective CRD
objects and modify the IP Tables as needed on their host. An alternative to this approach could be to add all of the
datapath functionality needed for globalnet to the RouteAgent, however this would result in a much higher level of
complexity for the RA, and a bigger re-write.

> #### Expose ClusterGlobalEgressIPs to Remote Clusters

Reviewer 5: ```suggestion
#### Expose ClusterGlobalEgressIPs, GlobalEgressIPs and GlobalIngressIPs to Remote Clusters

**NOTE:** In the PoC, only the `clusterglobalegressips` CRD Object was considered as described below. However,  `globalegressips` and `globalingressips` CRD Objects will need similar changes.

```

added to each remote GW as follows.

```shell
Routing Table 100:
<Remote GW 1 SNAT IP 1> via <Remote GW 1 IP> dev vxlan-tunnel metric 98
…
<Remote GW 1 SNAT IP X> via <Remote GW 1 IP> dev vxlan-tunnel metric 98
<Remote GW 2 SNAT IP 1> via <Remote GW 2 IP> dev vxlan-tunnel metric 98
…
<Remote GW 2 SNAT IP X> via <Remote GW 2 IP> dev vxlan-tunnel metric 98
```

## Scaling considerations

* Additional tunnels between clusters are required for this design.

## Notes and Assumptions

* The initial POC will use standard external IP addresses for gateways, but load balancer mode is
  a high priority for the final implementation.
* The initial design and POC is focused on CNIs that use the kubeproxy route agent handler that establishes VXLAN
  tunnels between worker nodes and gateways. Other options need to be investigated and supported.
* The multipath routes require an interface, so this approach won't work with IPsec Tunnel Mode.  IPsec Transport mode,
  proposed in [IPSec transport mode support](https://github.com/submariner-io/enhancements/pull/103) will be required for

>   IPsec encryption.

Reviewer 5: ```suggestion
  IPsec encryption.
* During the review of the Enhancement Proposal, there was concern raised regarding the SNAT that occurs in Step 3-b (See
   #multiple-active-gateway-packet-flow-without-globalnet and #multiple-active-gateway-packet-flow-with-globalnet).
   The concern is that up to this point, Submariner has preserved the source IP all the way to the destination pod.
   This may be needed in the future for policy enforcement. The SNAT is in place so the packet returns to the Gateway
   the packet came from within this cluster so that connection tracking can reverse the DNAT that occurred when the
   Service IP was resolved to a Pod IP. At this time, we do not have a work around. 
```

> ## TODO

Reviewer 5: ADD this section. You might get some linting errors. Make sure last section views correctly. There are multi-line bullets.

## PoC

To prove out the architecture described in this document, a PoC was implemented and can be found in:
* https://github.com/submariner-io/submariner/tree/feature-multi-active-gw 
* https://github.com/submariner-io/submariner-operator/tree/feature-multi-active-gw 

Additional details regarding the PoC can be found in the
[README.md](https://github.com/submariner-io/submariner/blob/feature-multi-active-gw/README.md) 
on the feature branch.

The PoC implements Multiple Active Gateways with both Globalnet and non-Globalnet deployments.
Use the following commands to deploy in a set of KIND clusters. The globalnet flags are optional.
Because there are now additional Gateways, additional nodes need the `submariner.io/gateway=true` label.

Using `make deploy`:

```
make deploy using=lighthouse,vxlan,globalnet DEPLOY_ARGS='--deploytool_submariner_args="--multi-active-gateway=true"'

kubectl config use-context cluster1
kubectl label nodes cluster1-worker2 submariner.io/gateway=true

kubectl config use-context cluster2
kubectl label nodes cluster1-worker2 submariner.io/gateway=true
```

Manually deploying:

```
make clusters

kubectl config use-context cluster1
kubectl label nodes cluster1-worker submariner.io/gateway=true
kubectl label nodes cluster1-worker2 submariner.io/gateway=true

subctl deploy-broker --kubeconfig output/kubeconfigs/kind-config-cluster1 --repository localhost:5000 --version local --globalnet

subctl join --kubeconfig output/kubeconfigs/kind-config-cluster1 broker-info.subm --clusterid cluster1 --natt=false --repository localhost:5000 --version local --cable-driver vxlan --multi-active-gateway=true

kubectl config use-context cluster2
kubectl label nodes cluster1-worker submariner.io/gateway=true
kubectl label nodes cluster1-worker2 submariner.io/gateway=true

subctl join --kubeconfig output/kubeconfigs/kind-config-cluster2 broker-info.subm --clusterid cluster2 --natt=false --repository localhost:5000 --version local --cable-driver vxlan --multi-active-gateway=true
 ```

This is a PoC and there were some shortcuts taken and there are some limitations.
* All nodes must be in Multiple Active Gateway mode.
  This won’t be required in the final implementation.
* Headless services are broken.
  Every Globalnet Controller was not rewritten until architecture was agreed upon.
* Only the VxLAN Cabledriver is supported. 
* Some of the `subctl` commands like `subctl diagnose` were not modified to take into account MAG.
  Unit Tests were not written for the new feature, some existing tests were modified to fix compilation
  issues or get existing tests to pass.
* The new `multi-active-gateway` flag was added to both implementations of `subctl` (old and new).
  In the time since the PoC was completed, `subctl` has been moved to its own repo.
  So this PoC code will need to be ported to the new repo.
  **Caution when rebasing to the feature-multi-active-gw branch to devel.**

* Performance testing on POC code.
* Support non-kubeproxy route agent handlers.
* Support Loadbalancer mode.

> * Support detection and handling of gateway failure and recovery.

Reviewer 5: ```suggestion
* Support detection and handling of gateway failure and recovery.
* In the PoC, only the new ClusterGlobalEgressIP Controller was implemented in the Gateway Pod (See #multiple-active-gateway-changes). The other controllers need to be ported from Globalnet.
```

```shell
Routing Table 100:
<Remote GW 1 SNAT IP 1> via <Remote GW 1 IP> dev vxlan-tunnel metric 98
…
<Remote GW 1 SNAT IP X> via <Remote GW 1 IP> dev vxlan-tunnel metric 98
<Remote GW 2 SNAT IP 1> via <Remote GW 2 IP> dev vxlan-tunnel metric 98
…
<Remote GW 2 SNAT IP X> via <Remote GW 2 IP> dev vxlan-tunnel metric 98
```

## Scaling considerations

* Additional tunnels between clusters are required for this design.

## Notes and Assumptions

* The initial POC will use standard external IP addresses for gateways, but load balancer mode is
  a high priority for the final implementation.
* The initial design and POC is focused on CNIs that use the kubeproxy route agent handler that establishes VXLAN
  tunnels between worker nodes and gateways. Other options need to be investigated and supported.
* The multipath routes require an interface, so this approach won't work with IPsec Tunnel Mode.  IPsec Transport mode,
  proposed in [IPSec transport mode support](https://github.com/submariner-io/enhancements/pull/103) will be required for
  IPsec encryption.
* During the review of the Enhancement Proposal, there was concern raised regarding the SNAT that occurs in Step 3-b (See

>    #multiple-active-gateway-packet-flow-without-globalnet and #multiple-active-gateway-packet-flow-with-globalnet).

Reviewer 5: My bad:
```suggestion
   [Multiple Active Gateway Packet Flow without Globalnet](#multiple-active-gateway-packet-flow-without-globalnet) and
   [Multiple Active Gateway Packet Flow with Globalnet](#multiple-active-gateway-packet-flow-with-globalnet)).
```

   The concern is that up to this point, Submariner has preserved the source IP all the way to the destination pod.
   This may be needed in the future for policy enforcement. The SNAT is in place so the packet returns to the Gateway
   the packet came from within this cluster so that connection tracking can reverse the DNAT that occurred when the
   Service IP was resolved to a Pod IP. At this time, we do not have a work around. 

## TODO

* Performance testing on POC code.
* Support non-kubeproxy route agent handlers.
* Support Loadbalancer mode.
* Support detection and handling of gateway failure and recovery.

> * In the PoC, only the new ClusterGlobalEgressIP Controller was implemented in the Gateway Pod (See #multiple-active-gateway-changes). The other controllers need to be ported from Globalnet.

Reviewer 5: ```suggestion
* In the PoC, only the new ClusterGlobalEgressIP Controller was implemented in the Gateway Pod (See [Multiple Active Gateway Changes](#multiple-active-gateway-changes)). The other controllers need to be ported from Globalnet.
```
